services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/utils/MongoDbCallExecutor.java

```diff
@@ -0,0 +1,152 @@
+        this.vertx = Objects.requireNonNull(vertx);
+        this.config = Objects.requireNonNull(config);
+        this.mongoClient = MongoClient.createShared(vertx, getMongoClientConfigAsJson());
+    }
+
+    /**
+     * Gets the mongo client.
+     *
+     * @return The mongo client.
+     */
+    public MongoClient getMongoClient() {
+        return mongoClient;
+    }
+
+    /**
+     * Creates mongodb collection index. Wrapper of {@link #createIndex(String, JsonObject, IndexOptions, Handler)}
+     *
+     * @param collectionName The name of the collection of documents.
+     * @param keys           The keys to be indexed.
+     * @param options        The options used to configure index, which is optional.
+     * @return  A future indicating the outcome of the indices creation operation.
+     */
+    public Future<Void> createCollectionIndex(final String collectionName, final JsonObject keys,
+                                              final IndexOptions options) {
+        final Promise<Void> indexCreationPromise = Promise.promise();
+        createIndex(collectionName, keys, options, res -> {
+            if (res.succeeded()) {
+                indexCreationPromise.complete();
+            } else if (res.cause() instanceof MongoSocketException) {
+                log.info("Create indices failed, wait for retry, cause:", res.cause());
+                vertx.setTimer(this.config.getCreateIndicesTimeout(),
+                        timer -> createIndex(collectionName, keys, options, res2 -> {
+                            if (res2.succeeded()) {
+                                indexCreationPromise.complete();
+                            } else if (res2.cause() instanceof MongoSocketException) {
+                                log.info("Create indices failed, wait for second retry, cause:", res2.cause());
+                                vertx.setTimer(this.config.getCreateIndicesTimeout(),
+                                        timer2 -> createIndex(collectionName, keys, options, res3 -> {
+                                            if (res3.succeeded()) {
+                                                indexCreationPromise.complete();
+                                            } else {
+                                                log.error("Error creating index", res3.cause());
+                                                indexCreationPromise.fail(res3.cause());
+                                            }
+                                        }));
+                            } else {
+                                log.error("Error creating index", res2.cause());
+                                indexCreationPromise.fail(res2.cause());
+                            }
+                        }));
+            } else {
+                log.error("Error creating index", res.cause());
+                indexCreationPromise.fail(res.cause());
+            }
+        });
```

Comment:
I believe this can be done more elegantly using recursion, can't it?

Reply:
This is the contribution code from the hub and the next step is making optimisations as you suggested.

Reply:
I assume you have made a note to your journal so that you don't forget it ;-)

Reply:
sure :-)

Reply:
@sophokles73 As we discussed above, I have now refactored the index creation code using recursion and pushed the changes.